### ISSUE TITLE
BSD support

### ISSUE THREAD

**Issue by bakkeby** [bakkeby/dusk#9]
There is currently no support for BSD in dusk because the IPC (inter-process communication) implementation uses [epoll](https://man7.org/linux/man-pages/man7/epoll.7.html) which is a very Linux centric facility for I/O event notification.

The IPC logic is used to allow the dusk client (duskc) to communicate with dusk to run commands, set the status, get information, etc.

One option would be to port the [IPC](https://dwm.suckless.org/patches/ipc/) patch ([github](https://github.com/mihirlad55/dwm-ipc)) to use the cross-platform [libevent](https://libevent.org/) library to handle the lower end communication (epoll, [kqueue](https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2), etc.). There is also a smaller [libev](http://software.schmorp.de/pkg/libev.html) library, but this is not so readily available on all platforms to my knowledge.

Alternatively one could use the `kqueue` library directly and have compile time flags to switch between the `kqueue` or `epoll` implementation. Dropping IPC altogether would be less than ideal.

This may be a waste of effort if there is no interest in support for BSD of course so add a 👍🏻 to this issue to show interest.

If you would like to get involved regarding this then that would be most appreciated.

**Comment by explosion-mental** [comment]
what about using `poll`? I had a bit of fun (annoyance) figuring out how to implement it in my build, but it is very straight foward.

**Comment by bakkeby** [comment]
https://www.freebsd.org/cgi/man.cgi?query=poll&manpath=FreeBSD+13.0-RELEASE+and+Ports
https://man7.org/linux/man-pages/man2/poll.2.html

Considering that the IPC patch only uses a single file descriptor there shouldn't be any real downsides to using traditional `poll` in this context.

This page (and associated videos) explain the difference between `poll`, `select` and `epoll`:
https://www.softprayog.in/programming/io-multiplexing-select-poll-epoll-in-linux

What is interesting on that page is that there is example code implemented in both `poll` and `epoll` so that they can be compared and it is primarily the monitoring code that is different.

**Comment by explosion-mental** [comment]
> Considering that the IPC patch only uses a single file descriptor there shouldn't be any real downsides to using traditional poll in this context.

Yes, but remember that `poll` can handle multiple fds. I think `epoll` has the ability to store some kind of data (i think i saw `64`, `32` and `16` for it) and that would be one major difference between plain `poll`, which only monitors file descriptors 'state' (for writting or reading).

While searching about this those links would been so useful.. What I did stumble across was that '`poll` is prefered over `select`'.

Might interest you an example, the [run](https://github.com/explosion-mental/Dwm/blob/37de553e8729461ffddf837acb533ee9b3a09460/dwm.c#L2937) function in my build it's all I do with `poll`.

**Comment by explosion-mental** [comment]
just throwing out that the cool kids now use `io_uring` :p

**Comment by bakkeby** [comment]
Note to self; have a closer look at how the IPC implementation in https://github.com/JLErvin/berry is done.

**Comment by bakkeby** [comment]
After moving to dbus for IPC instead of the IPC patch that uses a file socket for communication it wasn't much of a stretch to make IPC optional altogether.

So it should in principle be possible to get dusk running on BSD without the dusk client which isn't strictly needed anyway.

Perhaps it will make sense to maintain a separate config.mk for BSD (and maybe the Makefile as well).

**Comment by bakkeby** [comment]
I have successfully managed to run dusk on OpenBSD and I have added a guide with my notes on what I did to get it running.

https://github.com/bakkeby/dusk/wiki/How-to-set-up-dusk-for-BSD

I will need to run through this on FreeBSD as well to work out what additional changes are needed there.
